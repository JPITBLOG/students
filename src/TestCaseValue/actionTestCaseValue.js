export const getAllStudent = () => {
    return [{
        "_id": "5db15d02a234701cd4500d45",
        "fname": "Biren",
        "lname": "Pithavala",
        "dob": "2017-09-23",
        "studentImage": "http://localhost:8001\\uploads\\student\\2019-10-24T08-12-50.250Zacer1.jpeg",
        "address_one": "SultanaBad,Bhimpor,Surat",
        "address_two": "SultanaBad,Bhimpor,Surat",
        "state": "Gujarat",
        "city": "Surat",
        "zipcod": "343432",
        "isdelete": "0",
        "marks": [
            {
                "_id": "5db15d02a234701cd4500d46",
                "subjectId": "5dada92127bf5f0124f5aac0",
                "marks": "34",
                "isdelete": "0"
            },
            {
                "_id": "5db15d02a234701cd4500d47",
                "subjectId": "5dada8f927bf5f0124f5aabf",
                "marks": "36",
                "isdelete": "0"
            }
        ]
    },
        {
            "_id": "5dadafb927bf5f0124f5aac7",
            "fname": "Tejash",
            "lname": "Prajapati",
            "dob": "2018-09-13",
            "studentImage": "http://localhost:8001\\uploads\\student\\2019-10-21T13-16-41.396Zacer5.jpeg",
            "address_one": "Nanpura,Surat,Gujarat",
            "address_two": "Nanpura,Surat,Gujarat",
            "state": "Gujarat",
            "city": "Surat",
            "zipcod": "395007",
            "isdelete": "0",
            "marks": [
                {
                    "_id": "5dadafb927bf5f0124f5aac8",
                    "subjectId": "5dada94127bf5f0124f5aac2",
                    "marks": "30",
                    "isdelete": "0"
                },
                {
                    "_id": "5dadafb927bf5f0124f5aac9",
                    "subjectId": "5dada93527bf5f0124f5aac1",
                    "marks": "35",
                    "isdelete": "0"
                }
            ]
        }]
}

export const editData = () => {
    return [{
        '_id':"0123",
        'fname': "fname",
        'lname': "lname",
        'dob': "11-8-95",
        'studentImage': "there is image file",
        'address_one': "this is my first address",
        'address_two': "this is my first address",
        'state': "Gujarat",
        'city': "Surat",
        'zipcod': "122211",
        'isdelete': 0,
        'marks':[{_id:"1",subject:"Gujarati",marks:23,isdelete: 0},
            {_id:"2",subject:"Maths",marks:40,isdelete: 1}]
    }]
}

export const passSubject = () => {
    return [
        {
            "_id": "5dada8f927bf5f0124f5aabf",
            "subject": "Gujarati"
        },
        {
            "_id": "5dada92127bf5f0124f5aac0",
            "subject": "Hindi"
        },
        {
            "_id": "5dada93527bf5f0124f5aac1",
            "subject": "English"
        },
        {
            "_id": "5dada94127bf5f0124f5aac2",
            "subject": "Maths"
        }
    ]
}