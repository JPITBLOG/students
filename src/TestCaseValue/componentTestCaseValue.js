export const getAllStudent = () => {
    return [{
        _id: "0123",
        address_one: "SultanaBad,Bhimpor,Surat..",
        address_two: "SultanaBad,Bhimpor,Surat..",
        city: "Surat..",
        dob: "2017-09-23",
        studentImage: "http://localhost:8001\\uploads\\student\\2019-10-30T08-46-45.643Zasus1.jpeg",
        fname: "Biren",
        isdelete: "0",
        lname: "Pithavala",
        marks: [{
            isdelete: "0",
            marks: "34",
            subjectId: "5dada92127bf5f0124f5aac0",
            _id: "5db15d02a234701cd4500d46"
        }, {
            isdelete: "0",
            marks: "36",
            subjectId: "5dada8f927bf5f0124f5aabf",
            _id: "5db15d02a234701cd4500d47"
        }]
    },{
        _id: "01234",
        address_one: "SultanaBad,Bhimpor,Surat",
        address_two: "SultanaBad,Bhimpor,Surat",
        city: "Surat",
        dob: "2017-09-23",
        studentImage: "http://localhost:8001\\uploads\\student\\2019-10-30T08-46-45.643Zasus1.jpeg",
        fname: "Biren",
        isdelete: "0",
        lname: "Pithavala",
        marks: [{
            isdelete: "0",
            marks: "34",
            subjectId: "5dada92127bf5f0124f5aac0",
            _id: "5db15d02a234701cd4500d46"
        }, {
            isdelete: "0",
            marks: "36",
            subjectId: "5dada8f927bf5f0124f5aabf",
            _id: "5db15d02a234701cd4500d47"
        }]
    }]
}

export const subjectData = () => {
    return [{subject: "Gujarati",
            _id: "5dada8f927bf5f0124f5aabf"},
            {subject: "Hindi",
                _id: "5dada92127bf5f0124f5aac0"}]
}

export const studentData = () => {
    return {
        _id: "0123",
        address_one: "SultanaBad,Bhimpor,Surat..",
        address_two: "SultanaBad,Bhimpor,Surat..",
        city: "Surat..",
        dob: "2017-09-23",
        studentImage: "http://localhost:8001\\uploads\\student\\2019-10-30T08-46-45.643Zasus1.jpeg",
        fname: "Biren",
        isdelete: "0",
        lname: "Pithavala",
        marks: [{
        isdelete: "0",
        marks: "34",
        subjectId: "5dada92127bf5f0124f5aac0",
        _id: "5db15d02a234701cd4500d46"
    }, {
        isdelete: "0",
        marks: "36",
        subjectId: "5dada8f927bf5f0124f5aabf",
        _id: "5db15d02a234701cd4500d47"
    }]
    }
}

export const setEditData = () => {
    return {
        _id: "01234",
            address_one: "SultanaBad,Bhimpor,Surat..",
        address_two: "SultanaBad,Bhimpor,Surat..",
        city: "Surat..",
        dob: "2017-09-23",
        studentImage: "http://localhost:8001\\uploads\\student\\2019-10-30T08-46-45.643Zasus1.jpeg",
        fname: "Biren",
        isdelete: "0",
        lname: "Pithavala",
        marks: [{
        isdelete: "0",
        marks: "34",
        subjectId: "5dada92127bf5f0124f5aac0",
        _id: "5db15d02a234701cd4500d46"
    }, {
        isdelete: "0",
        marks: "36",
        subjectId: "5dada8f927bf5f0124f5aabf",
        _id: "5db15d02a234701cd4500d47"
    }]
    }
}

export const matchAddElementState = () => {
    return [{_id:"5db15d02a234701cd4500d46",subject:"5dada92127bf5f0124f5aac0",marks:"34",isdelete:0},
            {_id:"5db15d02a234701cd4500d47",subject:"5dada8f927bf5f0124f5aabf",marks:"36",isdelete:0},
            {_id:null,subject:null,marks:null,isdelete:0}]
}

export const marksDetailData = () => {
    return [{_id:"123",subject:"gujarati",marks:20,isdelete:1},{_id:"234",subject:"hindi",marks:32,isdelete:0},
        {_id:"343",subject:"maths",marks:25,isdelete:1}]
}

export const errorData = () => {
    return {fname: "there is an error in fname",lname: "there is an error in lname",dob: "there is an error in dob",
        selectedFile: "there is an error in select file",address1: "there is an error in address1",
        address2: "there is an error in address2",state: "there is an error in state",city: "there is an error in city",
        zipcod: "there is an error in zipcode"}
}

export const stateData = () => {
    return {fname:"jigar",lname:"patel",dob:'23/8/1995',srcImage:"http://localhost:8001\\uploads\\student\\2019-10-30T08-46-45.643Zasus1.jpeg",
        address1:"this is my first address",address2:"this is my second address",cstat:"Gujarat",city:"Surat",zipcod:"395007"}
}