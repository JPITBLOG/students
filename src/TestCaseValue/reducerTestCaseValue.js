export const stateValue = () => {
    return {Allstudent:[{
        "_id": "5db15d02a234701cd4500d45",
        "fname": "Biren",
        "lname": "Pithavala",
        "dob": "2017-09-23",
        "studentImage": "http://localhost:8001\\uploads\\student\\2019-10-24T08-12-50.250Zacer1.jpeg",
        "address_one": "SultanaBad,Bhimpor,Surat",
        "address_two": "SultanaBad,Bhimpor,Surat",
        "state": "Gujarat",
        "city": "Surat",
        "zipcod": "343432",
        "isdelete": "0",
        "marks": [
            {
                "_id": "5db15d02a234701cd4500d46",
                "subjectId": "5dada92127bf5f0124f5aac0",
                "marks": "34",
                "isdelete": "0"
            },
            {
                "_id": "5db15d02a234701cd4500d47",
                "subjectId": "5dada8f927bf5f0124f5aabf",
                "marks": "36",
                "isdelete": "0"
            }
        ]
    }]}
}

export const editValue = () => {
    return [{
        "_id": "5db15d02a234701cd4500d45",
        "fname": "Hiren",
        "lname": "Pithavala",
        "dob": "2017-09-21",
        "studentImage": "http://localhost:8001\\uploads\\student\\2019-10-24T08-12-50.250Zacer1.jpeg",
        "address_one": "SultanaBad,Bhimpor,Surat..",
        "address_two": "SultanaBad,Bhimpor,Surat..",
        "state": "Gujarat",
        "city": "Surat",
        "zipcod": "343432",
        "isdelete": "0",
        "marks": [
            {
                "_id": "5db15d02a234701cd4500d46",
                "subjectId": "5dada92127bf5f0124f5aac0",
                "marks": "34",
                "isdelete": "0"
            },
            {
                "_id": "5db15d02a234701cd4500d47",
                "subjectId": "5dada8f927bf5f0124f5aabf",
                "marks": "36",
                "isdelete": "0"
            }
        ]
    }]
}