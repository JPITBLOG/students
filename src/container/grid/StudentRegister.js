import React,{Component} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import StudentForm from './StudentForm';
import StudentMarks from './StudentMarks';
import classnames from 'classnames';
import {Button,
    Modal,
    ModalHeader,
    ModalBody,
    ModalFooter,
    TabContent,
    TabPane,
    Nav,
    NavItem,
    NavLink,
    Row, Col } from 'reactstrap';
import * as AddStudentDetailAction from "../../action/AddStudentDetail";
import * as getDataLengthAction from "../../action/dataLength";
import * as editStudentAction from "../../action/editStudentDetail";

export class StudentRegister extends Component{
    constructor(props){
        super(props);
        this.toggle = this.toggle.bind(this);
        this.onChangeHandler = this.onChangeHandler.bind(this);
        this.getEditedData = this.getEditedData.bind(this);
        
        this._id = null;
        this.fname = null;
        this.lname = null;
        this.dob = null;
        this.selectedFile = null;
        this.srcImage = null;
        this.address1 = null;
        this.address2 = null;
        this.cstat = null;
        this.city = null;
        this.zipcod = null;
        this.marksDetails = []

        this.state = {
            activeTab: '1',
            _id: null,
            fname: null,
            lname: null,
            dob: null,
            selectedFile: null,
            srcImage: null,
            address1: null,
            address2: null,
            cstat: null,
            city: null,
            zipcod: null,
            errors : {},
            error_flag:false,
            subject : [{_id:0,subject:"select"}],
            marksDetail : [{_id:null,subject:null,marks:null,isdelete:0}],
            deletedMarksDetail : [],
            selectinputs: ['selectinput-0'],
            inputs: ['input-0'],
            link: ['link-0']
        }
    }

    componentDidMount() {
        let subjectdata = this.props.getAllSubject.Allsubject;
        if(this.state.subject.length === 1){
            this.setState({subject:[...this.state.subject,...subjectdata]});
        }
    }

    componentDidUpdate = () => {
        this.getEditedData()
    }

    onChangeHandler = (event) => {
        if(event.target.files && event.target.files[0]){
            this.setState({
                selectedFile: event.target.files[0],
                srcImage: URL.createObjectURL(event.target.files[0])
            })
        }
    }

    handleChange = (e,stateKey) => {
        this.setState({[stateKey]:e.target.value})
    }

    onSubjectSelect = (e,index,name) => {
        let marksDetail = this.state.marksDetail.slice();
        if(e.target.value === "0") marksDetail[index][name] = null    
        else marksDetail[index][name] = e.target.value;
        this.setState({marksDetail:marksDetail})
    }

    //addNewElement
    addElement = () => {
        let newselectInput = `selectinput-${this.state.selectinputs.length}`;
        let newinputs = `input-${this.state.inputs.length}`;
        let newlink = `link-${this.state.link.length}`;

        this.setState(prevState => ({ selectinputs: prevState.selectinputs.concat([newselectInput]),
            inputs: prevState.inputs.concat([newinputs]),link: prevState.link.concat([newlink]),
            marksDetail: prevState.marksDetail.concat({_id:null,subject:null,marks:null,isdelete:0})}));
    }

    //delete element
    deleteElement = (event,index) => {
        if(this.state.selectinputs.length > 1){
            let deletedMarksDetail;
            let  selectinput = this.state.selectinputs.slice();
            let inputs = this.state.inputs.slice();
            let link=  this.state.link.slice();
            let marksDetail = this.state.marksDetail.slice();
            selectinput.splice(index,1);
            inputs.splice(index,1);
            link.splice(index,1);
            if(this.state._id !== null){
                if(marksDetail[index]._id !== null){
                    deletedMarksDetail = {_id:marksDetail[index]._id,subject:marksDetail[index].subject,
                        marks:marksDetail[index].marks,isdelete:1}
                }
            }
            marksDetail.splice(index,1);
            if(deletedMarksDetail !== undefined)
                this.setState({selectinputs:selectinput,inputs:inputs,link:link,marksDetail:marksDetail,deletedMarksDetail:[...this.state.deletedMarksDetail,deletedMarksDetail]}) 
            else
                this.setState({selectinputs:selectinput,inputs:inputs,link:link,marksDetail:marksDetail})
        }
    }

    commonvalidation = (inputparameter) => {
        if(inputparameter == null || inputparameter.length < 15){
            if(inputparameter == null){
                return("Cannot be empty");
            }
            else if(inputparameter.length < 15){
                return("required atleast 15 character");
            }
            else if(inputparameter.length > 100){
                return("maximum length is 100 character");
            }
        }
        return true;
    }

    nameValidation = (nameparameter) => {
        //name
        let nameFormat = /^[A-Za-z]+$/;
        if(nameparameter !== null && nameparameter !== ''){
            if(!nameparameter.match(nameFormat)){
                return("only letter is allowed");
            }
            else if(nameparameter.length < 3){
                return("required atleast 3 character");
            }
            else if(nameparameter.length > 15){
                return("Maximum Length is 15 character.");
            }
        }
        else {
            return("Cannot be empty");
        }
        return true;
    }

    //validation
    handleValidation = () => {
        let errors = {};
        let commonresp;
        let formIsValidate = true;
        this._id = this.state._id;
        this.fname = this.state.fname;
        this.lname = this.state.lname;
        this.dob = this.state.dob;
        this.selectedFile = this.state.selectedFile;
        this.address1 = this.state.address1;
        this.address2 = this.state.address2;
        this.cstat = this.state.cstat;
        this.city = this.state.city;
        this.zipcod = this.state.zipcod;
        this.marksDetails = this.state.marksDetail;

        //fname
        commonresp = this.nameValidation(this.fname);
        if(commonresp !== true){
            errors['fname'] = commonresp;
            formIsValidate = false;
        }

        //lname
        commonresp = this.nameValidation(this.lname);
        if(commonresp !== true){
            errors['lname'] = commonresp;
            formIsValidate = false;
        }

        //dob
        if(this.dob === null || this.dob === ''){
            errors['dob'] = "Select Date of Birth in Proper Format";
            formIsValidate = false;
        }

        //image validation
        if(this.state._id === null){
            if(this.selectedFile){

                let extension = (this.selectedFile.name).substr((this.selectedFile.name).lastIndexOf('.')).toLowerCase();
                if(extension === ".jpg" || extension === ".jpeg" || extension === ".bmp" ||
                    extension === ".gif" || extension === ".png"){
                }
                else {
                    errors["selectedFile"] = "Select only image file";
                    formIsValidate = false;
                }
            }
            else {
                errors["selectedFile"] = "Select image file";
                formIsValidate = false;
            }
        }

        //address1
        commonresp = this.commonvalidation(this.address1);
        if(commonresp !== true){
            errors['address1'] = commonresp;
            formIsValidate = false;
        }

        //address2
        commonresp = this.commonvalidation(this.address2);
        if(commonresp !== true){
            errors['address2'] = commonresp;
            formIsValidate = false;
        }

        //state
        commonresp = this.nameValidation(this.cstat);
        if(commonresp !== true){
            errors['state'] = commonresp;
            formIsValidate = false;
        }

        //city
        commonresp = this.nameValidation(this.city);
        if(commonresp !== true){
            errors['city'] = commonresp;
            formIsValidate = false;
        }

        //zipcod
        if(this.zipcod !== null && this.zipcod !== ''){
            let NumbarFormat = /^[0-9]*$/;
            if(!this.zipcod.match(NumbarFormat)){
                errors['zipcod'] = "only number is allowed";
                formIsValidate = false;
            }
            else if(this.zipcod.length < 6 || this.zipcod.length > 6 ){
                errors['zipcod'] = "6 numer is valid for zipcod";
                formIsValidate = false;
            }
        }
        else {
            errors['zipcod'] = "can't be empty";
            formIsValidate = false;
        }
        
        //marksDetail
        if(this.marksDetails !== undefined){
            let NumbarFormat = /^[0-9]*$/;
            this.marksDetails.map(function (marksDetail,index) {
                if(marksDetail.subject === null || marksDetail.marks === null || marksDetail.marks.length === 0){
                    if(marksDetail.subject === null){
                        errors['subject'+index] = "select subject at row: "+(index+1);
                        formIsValidate = false;
                    }
                    if(marksDetail.marks === null || marksDetail.marks.length === 0){
                        errors['marks'+index] = "enter marks at row: "+(index+1);
                        formIsValidate = false;
                    }
                }
                else if(!marksDetail.marks.match(NumbarFormat)){
                    errors['marks'+index] = "enter only Digit in Mark at: "+(index+1);
                    formIsValidate = false;
                }
                else {
                    if(marksDetail.marks.length > 3){
                        errors['marks'+index] = "Maximum 3 digit allow in Mark at: "+(index+1);
                        formIsValidate = false;
                    }
                }
                return 0;
            })
        }
        else {
            errors['subject'] = "enter student marks detail";
            formIsValidate = false;
        }

        this.setState({errors : errors});
        return formIsValidate;
    }

    setEditedData = (student) => {
        let selectinputs = [];
        let input = [];
        let link = [];
        let marksDetail = [];
        if(this.state._id === null || this.state._id !== student._id){
            student.marks.map(function (mark,index) {
                selectinputs.push(`selectinput-${index}`);
                input.push(`input-${index}`);
                link.push(`link-${index}`);
                marksDetail.push({_id:mark._id,subject:mark.subjectId,marks:mark.marks,isdelete:0});
                return 0;
            });
            let studentImage = student.studentImage.split(':');
            studentImage = studentImage[3];
            this.setState({
                activeTab: '1',
                _id :student._id,
                fname :student.fname,
                lname :student.lname,
                dob: student.dob,
                selectedFile: studentImage,
                srcImage: student.studentImage,
                address1: student.address_one,
                address2: student.address_two,
                cstat: student.state,
                city: student.city,
                zipcod: student.zipcod,
                marksDetail : marksDetail,
                selectinputs: selectinputs,
                inputs: input,
                link: link,
                errors : {},
                error_flag:false
            });
        }
    }

    getEditedData = () => {
        if(Object.keys(this.props.studentData).length > 0) {
            let student = this.props.studentData;
            this.setEditedData(student);
        }
        else {
            if(this.state._id  !== null){
                this.clearState();
            }
        }
    }

    clearState = () => {
        this.setState({
            activeTab: '1',
            _id: null,
            fname: null,
            lname: null,
            dob: null,
            selectedFile: null,
            srcImage: null,
            address1: null,
            address2: null,
            cstat: null,
            city: null,
            zipcod: null,
            errors : {},
            error_flag:false,
            subject : [{_id:"0",subject:"select"}],
            marksDetail : [{_id:null,subject:null,marks:null,isdelete: 0}],
            selectinputs: ['selectinput-0'],
            inputs: ['input-0'],
            link: ['link-0']
        },
            () => {
                console.log(this.state);
            })
        return true;
    }

    registerdata = async () => {
        let marksDetail;
        if(this.handleValidation()){
            const data = new FormData();
            data.append('_id',this._id);
            data.append('fname', this.fname);
            data.append('lname', this.lname);
            data.append('dob', this.dob);
            data.append('studentImage', this.selectedFile);
            data.append('address_one', this.address1);
            data.append('address_two', this.address2);
            data.append('state', this.cstat);
            data.append('city', this.city);
            data.append('zipcod', this.zipcod);
            data.append('isdelete', 0);
        
            if(this.state._id !== null){
                if(this.state.deletedMarksDetail.length > 0){
                    marksDetail = [...this.marksDetails,...this.state.deletedMarksDetail]
                }
                else marksDetail = this.marksDetails;
            }
            else marksDetail = this.marksDetails;
            data.append('subjectContent',JSON.stringify(marksDetail))
            if(this.state._id !== null){
                this.props.action.editStudentAction.EditStudentDetail(data).then((editActionresp) => {
                    this.clearState();
                    this.props.toggle();
                })
            }
            else {
                this.props.action.AddStudentDetailAction.AddStudentDetail(data).then(async (addActionresp) => {
                    await this.clearState();
                    this.props.toggle();
                    if(addActionresp){
                        this.props.action.getDataLength.dataLength();
                    }
                })
            }
        }
    }

    toggle(tab) {
        if (this.state.activeTab !== tab) {
            this.setState({
                activeTab: tab
            });
        }
    }

    setTogalData = () => {
        this.clearState()
        this.props.toggle();
    }

    render (){
        return(
            <React.Fragment>
            <Modal isOpen={this.props.isOpen} toggle={this.setTogalData.bind(this)} className={this.props.className}>
                <ModalHeader toggle={this.setTogalData.bind(this)}>Student Data!</ModalHeader>
                <ModalBody>
                    <div>
                        <Nav tabs>
                            <NavItem>
                                <NavLink
                                    className={classnames({ active: this.state.activeTab === '1' })}
                                    onClick={() => { this.toggle('1'); }}
                                >
                                    Student Detail
                                </NavLink>
                            </NavItem>
                            <NavItem>
                                <NavLink
                                    className={classnames({ active: this.state.activeTab === '2' })}
                                    onClick={() => { this.toggle('2'); }}
                                >
                                    Marks Detail
                                </NavLink>
                            </NavItem>
                        </Nav>
                        <div ></div>
                        <TabContent activeTab={this.state.activeTab}>
                            <TabPane tabId="1">
                                <StudentForm error_flag = {this.state.error_flag} handleChange = {(event, name) => this.handleChange(event, name)}
                                             onChangeHandler = {(event) => this.onChangeHandler(event)} errors = {this.state.errors} state = {this.state}
                                />
                            </TabPane>
                            <TabPane tabId="2">
                                <Row>
                                    <Col sm="12">
                                        <StudentMarks onSubjectSelect = {(event,index,name) => this.onSubjectSelect(event,index,name)}
                                                      addElement = {this.addElement.bind(this)} deleteElement = {(event,index) => this.deleteElement(event,index)}
                                                      selectinputs = {this.state.selectinputs} subject = {this.state.subject} inputs = {this.state.inputs}
                                                      link = {this.state.link} marksDetail = {this.state.marksDetail} errors = {this.state.errors}/>
                                    </Col>
                                </Row>
                            </TabPane>
                        </TabContent>
                    </div>
                </ModalBody>
                <ModalFooter>
                    <Button className="btn-primary" onClick={this.registerdata.bind(this)}>{this.state._id !== null ? "Edit Data" : "Add Data"}</Button>
                </ModalFooter>
            </Modal>
            </React.Fragment>
        );
    }
}

const mapStateToProps = (state) => {
    const {getAllSubject}  = state;
    return {
        getAllSubject : getAllSubject
    }
};

const mapDispatchToProps = dispatch => ({
    action: {
        AddStudentDetailAction: bindActionCreators(AddStudentDetailAction,dispatch),
        getDataLength: bindActionCreators(getDataLengthAction, dispatch),
        editStudentAction: bindActionCreators(editStudentAction, dispatch)
    }
})

export default connect(mapStateToProps,mapDispatchToProps)(StudentRegister);