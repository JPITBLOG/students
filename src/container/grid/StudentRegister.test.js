import React from 'react';
import {StudentRegister} from './StudentRegister';
import {shallow, configure} from 'enzyme';
import {subjectData, studentData, setEditData, matchAddElementState} from '../../TestCaseValue/componentTestCaseValue';
import renderer from 'react-test-renderer';
import Adapter from "enzyme-adapter-react-16";

configure({adapter: new Adapter()});

const props = {
    getAllSubject:{Allsubject:subjectData()},
    studentData: studentData(),
    toggle: jest.fn(),
    isOpen: false
}

const file = ({url:'example.png',type: 'image/png'});

const e = {
    target: { value: 'jigar',files:[file]}
};

describe('Test case for Student Register',() => {
    let store;
    const student = setEditData();
    const mockPropSetToggle = jest.fn();
    global.URL.createObjectURL = jest.fn();
    const TogalDataSpy = jest.spyOn(StudentRegister.prototype, 'toggle');
    const setTogalDataSpy = jest.spyOn(StudentRegister.prototype, 'setTogalData');
    // const sethandleChange = jest.spyOn(StudentRegister.prototype, 'handleChange');

    let wrapper = shallow(<StudentRegister  {...props}/>);

    test('test studentRegister component',() => {
        expect(wrapper.state('subject')).toEqual([{_id:0,subject:"select"},...subjectData()])
    })
    test('test for addElement function',() => {
        wrapper.instance().addElement();
        expect(wrapper.state('selectinputs')).toEqual(['selectinput-0','selectinput-1','selectinput-2']);
        expect(wrapper.state('inputs')).toEqual(['input-0','input-1','input-2']);
        expect(wrapper.state('marksDetail')).toEqual(matchAddElementState())
    })
    test('test for deleteElement function',() => {
        wrapper.instance().deleteElement(this,1);
        expect(wrapper.state('selectinputs')).toEqual(['selectinput-0','selectinput-2']);
        expect(wrapper.state('inputs')).toEqual(['input-0','input-2']);
        expect(wrapper.state('link')).toEqual(['link-0','link-2']);
        expect(wrapper.state('deletedMarksDetail')).toEqual([{_id:"5db15d02a234701cd4500d47",subject:"5dada8f927bf5f0124f5aabf",marks:"36",isdelete:1}])
    })
    test('test for commonValidation function',() => {
        const commonValidationResp = wrapper.instance().commonvalidation('commonValidationParameter');
        expect(commonValidationResp).toBe(true);
    })
    test('test for nameValidation function',() => {
        const nameValidationResp = wrapper.instance().nameValidation('jigar12');
        expect(nameValidationResp).toBe("only letter is allowed");
    })
    test('test for setEditedData function',() =>{
        wrapper.instance().setEditedData(student);
        expect(wrapper.state('_id')).toEqual("0123");
    })
    test('test for clearState function',() => {
       wrapper.instance().clearState();
       expect(wrapper.state('_id')).toEqual("0123");
       expect(wrapper.state('fname')).toEqual("Biren");
    });
    test('test for toggle function',() => {
        wrapper.instance().toggle(2);
        expect(wrapper.state('activeTab')).toEqual(2);
    })
    test('test for render snapShot',() => {
            const Modal = wrapper.find('#modal');
            expect(Modal.prop('isOpen')).toEqual(false);
            wrapper.find("#navLink1").simulate('click');
            expect(TogalDataSpy).toHaveBeenCalledTimes(2);
            wrapper.instance().setTogalData();
            expect(setTogalDataSpy).toHaveBeenCalledTimes(1);
            expect(Modal.find('TabContent').prop('activeTab')).toEqual(2);
            expect(Modal.find('StudentForm').prop('error_flag')).toEqual(false);
            wrapper.instance().handleChange(e,'fname');
            expect(wrapper.state('fname')).toEqual('jigar');
            global.URL.createObjectURL = jest.fn(() => e.target.files[0]);
            wrapper.instance().onChangeHandler(e);
            expect(wrapper.state('srcImage')).toEqual({url:'example.png',type: 'image/png'});
            wrapper.setState({errors:{test:'test case error'}});
            expect(Modal.find('StudentForm').prop('errors')).toEqual({});
            wrapper.update();
            const component = renderer.create(
                <StudentRegister  {...props}/>
            );
            let componentsnap = component.toJSON();
            expect(componentsnap).toMatchSnapshot();
        })

})
