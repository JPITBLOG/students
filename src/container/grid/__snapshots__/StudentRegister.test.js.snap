// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test case for Student Register test for render snapShot 1`] = `null`;
